pkg/filesystem/local_directory_windows.go

```diff
@@ -0,0 +1,1043 @@
+
+func (d *localDirectory) OpenRead(name path.Component) (FileReader, error) {
+	f, err := d.open(name, DontCreate, os.O_RDONLY)
+	if err != nil {
+		return nil, err
+	}
+	return localFileReadWriter{File: f}, nil
+}
+
+func (d *localDirectory) OpenReadWrite(name path.Component, creationMode CreationMode) (FileReadWriter, error) {
+	f, err := d.open(name, creationMode, os.O_RDWR)
+	if err != nil {
+		return nil, err
+	}
+	return localFileReadWriter{File: f}, nil
+}
+
+func (d *localDirectory) OpenWrite(name path.Component, creationMode CreationMode) (FileWriter, error) {
+	f, err := d.open(name, creationMode, os.O_WRONLY)
+	if err != nil {
+		return nil, err
+	}
+	return localFileReadWriter{File: f}, nil
+}
+
+func (d *localDirectory) Link(oldName path.Component, newDirectory Directory, newName path.Component) error {
+	defer runtime.KeepAlive(d)
+
+	return newDirectory.Apply(localDirectoryLink{
+		oldHandle: d.handle,
+		oldName:   oldName,
+		newName:   newName,
+	})
+}
+
+func (d *localDirectory) Clonefile(oldName path.Component, newDirectory Directory, newName path.Component) error {
+	return status.Error(codes.Unimplemented, "Clonefile is only supported on Darwin")
+}
+
+func (d *localDirectory) lstat(name path.Component) (FileType, error) {
+	var handle windows.Handle
+	fileName := name.String()
+	err := ntCreateFile(&handle, windowsext.FILE_READ_ATTRIBUTES, d.handle, fileName, windows.FILE_OPEN, windows.FILE_OPEN_REPARSE_POINT)
+	if err != nil {
+		return FileTypeOther, err
+	}
+	var fileInfo windows.ByHandleFileInformation
+	err = windows.GetFileInformationByHandle(handle, &fileInfo)
+	if err != nil {
+		return FileTypeOther, err
+	}
+	fileAttributes := fileInfo.FileAttributes
+	fileType := FileTypeOther
+	switch {
+	case (fileAttributes & windows.FILE_ATTRIBUTE_REPARSE_POINT) == windows.FILE_ATTRIBUTE_REPARSE_POINT:
+		fileType = FileTypeSymlink
+	case (fileAttributes & windows.FILE_ATTRIBUTE_DIRECTORY) == windows.FILE_ATTRIBUTE_DIRECTORY:
+		fileType = FileTypeDirectory
+	default:
+		fileType = FileTypeExecutableFile // assume all regular files are executable
+	}
+	return fileType, nil
+}
+
+func (d *localDirectory) Lstat(name path.Component) (FileInfo, error) {
+	defer runtime.KeepAlive(d)
+
+	fileType, err := d.lstat(name)
+	if err != nil {
+		return FileInfo{}, err
+	}
+	return NewFileInfo(name, fileType), nil
+}
+
+func (d *localDirectory) Mkdir(name path.Component, perm os.FileMode) error {
+	defer runtime.KeepAlive(d)
+
+	var handle windows.Handle
+	defer windows.CloseHandle(handle)
+	// perm is ignored like os.Mkdir on Windows
+	err := ntCreateFile(&handle, windows.FILE_LIST_DIRECTORY, d.handle, name.String(),
+		windows.FILE_CREATE, windows.FILE_DIRECTORY_FILE|windows.FILE_OPEN_REPARSE_POINT)
+	return err
+}
+
+func (d *localDirectory) Mknod(name path.Component, perm os.FileMode, dev int) error {
+	return status.Error(codes.Unimplemented, "Creation of device nodes is not supported on Windows")
+}
+
+type fileFullDirInfo struct {
```

Comment:
I suppose structs like these should also be moved into windowsext, as those are also candidates for upstreaming into golang.org/x/sys, right?

Reply:
I am more certain the consts and NtSetFileInformation will fit in x/sys, but not so sure about the structs. There are many win32 APIs that can take different struct types and an enum to indicate which type is passed (and sometimes another int for the size of the out buffer). In x/sys/windows, they have consts for the enums but use *byte for the struct passed in. I think it's sensible for them to do so because oftentimes the structs are really there just to tell you the offsets in a byte buffer and do not encapsulate, like when the last member of the struct is the first character of a C string and rest of the string should follow in memory..

But let me move them to windowsext, even if x/sys doesn't want them, it will look cleaner.